### ISSUE TITLE
Provide better error message when EditorForModel() would render an input tag without a name

### ISSUE THREAD

**Issue by rynowak** [aspnet/Mvc#2721]
This code throws an `ArgumentNullException`

```
@using Microsoft.AspNet.Http
@model List<IPostedFile>
<html>
<body>
    <section>
        <h2>HtmlHelpers</h2>
        @using (Html.BeginForm(FormMethod.Post, htmlAttributes: new { enctype = "multipart/form-data" }))
        {
            <p>@Html.EditorForModel()</p>
            <p><input type="submit" /></p>
        }
</body>
</html>
```

You can see similar things when the top-level model is a `string` or a numeric type.

You can can also repro this with taghelpers:
```
        <h2>TagHelpers</h2>
        <asp:form enctype="multipart/form-data" asp-action="Home">
            <p><asp:input asp-for="@Model"/></p>
            <p><input type="submit" /></p>
        </asp:form>
```

For HTML helpers you can work around this using:
```
@Html.EditorForModel(templateName: null, htmlFieldName: "files"
```

No known workaround for the TagHelper case

Callstack
-----------
```
at Microsoft.AspNet.Mvc.Rendering.DefaultHtmlGenerator.GenerateInput(ViewContext viewContext, InputType inputType, ModelExplorer modelExplorer, String expression, Object value, Boolean useViewData, Boolean isChecked, Boolean setId, Boolean isExplicitValue, String format, IDictionary<String, Object> htmlAttributes)
at Microsoft.AspNet.Mvc.Rendering.DefaultHtmlGenerator.GenerateTextBox(ViewContext viewContext, ModelExplorer modelExplorer, String expression, Object value, String format, Object htmlAttributes)
at Microsoft.AspNet.Mvc.Rendering.HtmlHelper.GenerateTextBox(ModelExplorer modelExplorer, String expression, Object value, String format, Object htmlAttributes)
at Microsoft.AspNet.Mvc.Rendering.HtmlHelper.TextBox(String expression, Object value, String format, Object htmlAttributes) 
at Microsoft.AspNet.Mvc.Rendering.DefaultEditorTemplates.GenerateTextBox(IHtmlHelper htmlHelper, Object value, Object htmlAttributes)
at Microsoft.AspNet.Mvc.Rendering.DefaultEditorTemplates.GenerateTextBox(IHtmlHelper htmlHelper, String inputType, Object value)
at Microsoft.AspNet.Mvc.Rendering.DefaultEditorTemplates.GenerateTextBox(IHtmlHelper htmlHelper, String inputType) 
at Microsoft.AspNet.Mvc.Rendering.DefaultEditorTemplates.StringTemplate(IHtmlHelper htmlHelper) 
at Microsoft.AspNet.Mvc.Rendering.Internal.TemplateRenderer.Render() 
at Microsoft.AspNet.Mvc.Rendering.Internal.TemplateBuilder.Build() 
at Microsoft.AspNet.Mvc.Rendering.HtmlHelper.GenerateEditor(ModelExplorer modelExplorer, String htmlFieldName, String templateName, Object additionalViewData)
at Microsoft.AspNet.Mvc.Rendering.HtmlHelper.Editor(String expression, String templateName, String htmlFieldName, Object additionalViewData)
at Microsoft.AspNet.Mvc.Rendering.HtmlHelperEditorExtensions.EditorForModel(IHtmlHelper htmlHelper) 
```

**Comment by dougbu** [comment]
The `ArgumentNullException` is expected for this case. Otherwise the helper would generate a field with an empty `name` attribute and that's not legal.

Could improve the exception's `Message` and perhaps throw an `InvalidOperationException` instead. Could also get the `<input>` tag helper to notice when the user assigns a value to the `name` attribute, avoiding the empty field problem.

**Comment by danroth27** [comment]
@rynowak What does `IPostedFile` look like?

**Comment by danroth27** [comment]
@rynowak Also, for the `IPostedFile` case what does your action look like that is handling the form post?

**Comment by rynowak** [comment]
> @rynowak What does IPostedFile look like?

It's `Microsoft.AspNet.Http.IPostedFile` -- I'm sure you can find it if you look :laughing: .

You can repro this with a bunch of simple types. The key to the repro is that we turn `IPostedFile` into an `<input type="file" />` instead of rendering a bunch of individual fields.

There's no 'name' for the top-level `<input />` tag because it represents the whole model. In this case the error message is bad, it needs to tell you that you need to explicitly provide a name.

Action looks like:

```
[HttpPost]
public IActionResult Index(IPostedFile file)
{
}
```

**Comment by kulmugdha** [comment]
Checked in fix a6a7903b846eadd62f43866beedea826e75e2b3f
